pkg/controller/nodeipam/ipam/multi_cidr_range_allocator.go

```diff
@@ -0,0 +1,1105 @@
+			//   (cidr is marked as used),
+			// - already processed a Node successfully and allocated CIDRs for it
+			//   (cidr is marked as used),
+			// - already processed a Node but we did saw a "timeout" response and
+			//   request eventually got through in this case we haven't released
+			//   the allocated CIDRs (cidr is still marked as used).
+			// There's a possible error here:
+			// - NC sees a new Node and assigns CIDRs X,Y.. to it,
+			// - Update Node call fails with a timeout,
+			// - Node is updated by some other component, NC sees an update and
+			//   assigns CIDRs A,B.. to the Node,
+			// - Both CIDR X,Y.. and CIDR A,B.. are marked as used in the local cache,
+			//   even though Node sees only CIDR A,B..
+			// The problem here is that in in-memory cache we see CIDR X,Y.. as marked,
+			// which prevents it from being assigned to any new node. The cluster
+			// state is correct.
+			// Restart of NC fixes the issue.
+			if len(newNode.Spec.PodCIDRs) == 0 {
+				return ra.AllocateOrOccupyCIDR(newNode)
+			}
+			return nil
+		}),
+		DeleteFunc: controllerutil.CreateDeleteNodeHandler(ra.ReleaseCIDR),
+	})
+
+	return ra, nil
+}
+
+func (r *multiCIDRRangeAllocator) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	klog.Infof("Starting Multi CIDR Range allocator")
+	defer klog.Infof("Shutting down Multi CIDR Range allocator")
+
+	if !cache.WaitForNamedCacheSync("multi_cidr_range_allocator", stopCh, r.nodesSynced, r.clusterCIDRConfigSynced) {
+		return
+	}
+
+	for i := 0; i < cidrUpdateWorkers; i++ {
+		go r.worker(stopCh)
+		go r.cccWorker()
+	}
+
+	<-stopCh
+}
+
+func (r *multiCIDRRangeAllocator) worker(stopChan <-chan struct{}) {
+	for {
+		select {
+		case workItem, ok := <-r.nodeCIDRUpdateChannel:
+			if !ok {
+				klog.Warning("Channel nodeCIDRUpdateChannel was unexpectedly closed")
+				return
+			}
+			if err := r.updateCIDRsAllocation(workItem); err != nil {
+				// Requeue the failed node for update again.
+				r.nodeCIDRUpdateChannel <- workItem
+			}
+		case <-stopChan:
+			return
+		}
+	}
+}
+
+func (r *multiCIDRRangeAllocator) cccWorker() {
+	for r.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem deals with one key off the queue.  It returns false when it's time to quit.
+func (r *multiCIDRRangeAllocator) processNextWorkItem() bool {
+	cKey, quit := r.queue.Get()
+	if quit {
+		return false
+	}
+	defer r.queue.Done(cKey)
+
+	err := r.syncClusterCIDRConfig(cKey.(string))
+	r.handleErr(err, cKey)
+
+	return true
+}
+
+func (r *multiCIDRRangeAllocator) handleErr(err error, key interface{}) {
+	if err == nil {
+		r.queue.Forget(key)
+		return
+	}
+
+	if r.queue.NumRequeues(key) < maxRetries {
+		klog.Warningf("Error syncing clusterCIDRConfig %q, retrying. Error: %v", key, err)
+		r.queue.AddRateLimited(key)
+		return
+	}
+
+	klog.Warningf("Retry budget exceeded, dropping clusterCIDRConfig %q out of the queue: %v", key, err)
```

Comment:
"Error syncing ClusterCIDRConfig %q, not retrying as retry budget (%d retries) has been exceeded.

Reply:
Done